source/fe/fe_enriched.cc

```diff
@@ -1048,6 +1049,451 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+      // Now a call to color_enrichment[1](cell) should in turn call
+      // enrichments[4](cell). That is just what this lambda is doing.
+      color_enrichments.resize(num_colors);
+      for (unsigned int i = 0; i < num_colors; ++i)
+        {
+          color_enrichments[i] =
+            [&, i](const typename Triangulation<dim, spacedim>::cell_iterator
+                     &cell) {
+              unsigned int id = cell->material_id();
+
+              /*
+               * i'th color_enrichment function corresponds to (i+1)'th color.
+               * Since FE_Enriched takes function pointers, we return a
+               * function pointer.
+               */
+              return enrichments[cellwise_color_predicate_map.at(id).at(i + 1)]
+                .get();
+            };
+        }
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    make_fe_collection_from_colored_enrichments(
+      const unsigned int &                       num_colors,
+      const std::vector<std::set<unsigned int>> &fe_sets,
+      const std::vector<std::function<const Function<spacedim> *(
+        const typename Triangulation<dim, spacedim>::cell_iterator &)>>
+        &                              color_enrichments,
+      const FE_Q<dim, spacedim> &      fe_base,
+      const FE_Q<dim, spacedim> &      fe_enriched,
+      const FE_Nothing<dim, spacedim> &fe_nothing,
+      hp::FECollection<dim, spacedim> &fe_collection)
+    {
+      // define dummy function which is associated with FE_Nothing
+      using cell_function = std::function<const Function<spacedim> *(
+        const typename Triangulation<dim, spacedim>::cell_iterator &)>;
+      cell_function dummy_function;
+      dummy_function =
+        [=](const typename Triangulation<dim, spacedim>::cell_iterator &)
+        -> const Function<spacedim> * {
+        AssertThrow(false,
+                    ExcMessage("Called enrichment function for FE_Nothing"));
+        return nullptr;
+      };
+
+      using EnrichmentFunctions_2DVector =
+        std::vector<std::vector<std::function<const Function<spacedim> *(
+          const typename Triangulation<dim, spacedim>::cell_iterator &)>>>;
+
+      // loop through color sets and create FE_enriched element for each
+      // of them provided before calling this function, we have color
+      // enrichment function associated with each color.
+      for (unsigned int color_set_id = 0; color_set_id != fe_sets.size();
+           ++color_set_id)
+        {
+          std::vector<const FiniteElement<dim, spacedim> *> vec_fe_enriched(
+            num_colors, &fe_nothing);
+          EnrichmentFunctions_2DVector functions(num_colors, {dummy_function});
+
+          for (auto it = fe_sets[color_set_id].begin();
+               it != fe_sets[color_set_id].end();
+               ++it)
+            {
+              // Given a color id ( = *it), corresponding color enrichment
+              // function is at index id-1 because color_enrichments is
```

Comment:
the same question with that +1/-1 shift, what's the origin?

Reply:
If i want color function corresponding to an index i (say = 3), the corresponding color function is (i-1) i.e 2.